esptool.py

```diff
@@ -58,44 +63,36 @@ class ESPROM:
     ESP_OTP_MAC1    = 0x3ff00054
     ESP_OTP_MAC3    = 0x3ff0005c
 
-    # Sflash stub: an assembly routine to read from spi flash and send to host
-    SFLASH_STUB     = "\x80\x3c\x00\x40\x1c\x4b\x00\x40\x21\x11\x00\x40\x00\x80" \
-                      "\xfe\x3f\xc1\xfb\xff\xd1\xf8\xff\x2d\x0d\x31\xfd\xff\x41\xf7\xff\x4a" \
-                      "\xdd\x51\xf9\xff\xc0\x05\x00\x21\xf9\xff\x31\xf3\xff\x41\xf5\xff\xc0" \
-                      "\x04\x00\x0b\xcc\x56\xec\xfd\x06\xff\xff\x00\x00"
-
     def __init__(self, port=0, baud=ESP_ROM_BAUD):
         self._port = serial.Serial(port)
         # setting baud rate in a separate step is a workaround for
         # CH341 driver on some Linux versions (this opens at 9600 then
         # sets), shouldn't matter for other platforms/drivers. See
         # https://github.com/themadinventor/esptool/issues/44#issuecomment-107094446
         self._port.baudrate = baud
-        self.in_bootloader = False  # actually unknown, but assume not
 
-    """ Read bytes from the serial port while performing SLIP unescaping """
-    def read(self, length=1):
+    """ Read a SLIP packet from the serial port """
+    def read(self):
         b = ''
-        while len(b) < length:
-            d = self._port.read(length - len(b))
-            pos = 0
-            while pos < len(d):
-                c = d[pos]
-                pos = pos + 1
-                if c == '\xdb':
-                    if pos >= len(d):
-                        d = self._port.read(length - len(b))
-                        pos = 0
-                    c = d[pos]
-                    pos = pos + 1
-                    if c == '\xdc':
-                        b = b + '\xc0'
-                    elif c == '\xdd':
-                        b = b + '\xdb'
-                    else:
-                        raise FatalError('Invalid SLIP escape')
+        while True:
```

Comment:
so, it's tricky. we don't know upfront how much data stub is going to send, and pyserial (at least the posix implementation) is bent on reading at least as many bytes as instructed.
even timeout stops working once at least one byte has been read.
what they do there (in PosixPollSerial.read) doesn't make sense to me and seems like a bug, but it is what it is.

however, since with this change verification does not require reading out all the bytes and performs on-device digest, the impact of going back to reading byte by byte is reduced substantially.